docs: ipns (#184)
ipfs/specs: naming/README.md

# IPNS - Inter-Planetary Naming System

Authors:

  - Vasco Santos ([@vasco-santos](https://github.com/vasco-santos))
  
Reviewers:

  - Steven Allen ([@Stebalien](https://github.com/Stebalien))

-----
  
# Abstract

IPFS is powered by content-addressed data, which by nature is immutable: changing an object would change its hash, and consequently its address, making it a different object altogether. However, there are several use cases where we benefit from having mutable data. This is where IPNS gets into the equation.

All things considered, the IPFS naming layer is responsible for the creation of:

  - mutable pointers to objects
  - human-readable names

# Status of this spec

![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)

# Organization of this document

  - [Introduction](#introduction)
  - [IPNS Record](#ipns-record)
  - [Protocol](#protocol)
  - [Overview](#overview)
  - [API Spec](#api-spec)
  - [Integration with IPFS](#integration-with-ipfs)

# Introduction

> Each time a file added to IPFS needs to be modified, its content addressable idendifier will be automatically modified. As a consequence, the address previously used for getting that file needs to be updated by who is using it. As this is not pratical, IPNS was conceptualized to solve the problem.

Reviewer: ```suggestion
Each time a file is modified, its content address changes. As a consequence, the address previously used for getting that file needs to be updated by who is using it. As this is not pratical, IPNS was created to solve the problem.
```

Wording nits:

* IPFS doesn't react to need, unfortunately.
* IPNS is more than a concept (it exists).

"needs to be updated by who is using it" is still awkward but I'm not sure of a better wording off the top of my head.

IPNS is based on [SFS](http://en.wikipedia.org/wiki/Self-certifying_File_System). It consists of a PKI namespace, where a name is simply the hash of a public key. As a result, whoever controls the private key has full control over the name. Accordingly, records are signed by the private key and then distributed across the network (in IPFS, via the routing system). This is an egalitarian way to assign mutable names on the Internet at large, without any centralization whatsoever, or certificate authorities.

# IPNS Record

> An IPNS record is a data structure composed by:

Reviewer: ```suggestion
An IPNS record is a data structure containing the following fields:
```

("composed by" isn't really idiomatic)

- 1. **Value** (bytes)

>   - It can be a path to another IPNS record, a `dnslink` path (`/ipns/example.com`) or an IPFS path (`/ipfs/Qm...`)

Reviewer: We should make it clear that these are *just* examples. It can be *any* path.

- 2. **Validity** (bytes)
  - Expiration date of the record using [RFC3339](https://www.ietf.org/rfc/rfc3339.txt) with nanoseconds precision.
  - Note: Currently, the expiration date is the only available type of validity.
- 3. **Validity Type** (uint64)

>   - Allows us to define if the record is valid under certain conditions

Reviewer: ```suggestion
   - Allows us to define the conditions under which the record is valid.
```

>   - Only supporting expiration date with `validityType = 0` for now

Reviewer: ```suggestion
   - Only supports expiration date with `validityType = 0` for now.
```

IPNS was idealized based on [SFS](http://en.wikipedia.org/wiki/Self-certifying_File_System). It consists of a PKI namespace, where a name is simply the hash of a public key. As a result, whoever controls the private key has full control over the name. Accordingly, records are signed by the private key and then distributed across the network (in IPFS, via the routing system). This is an egalitarian way to assign mutable names on the Internet at large, without any centralization whatsoever, or certificate authorities.

# Building Blocks

## IPNS Record

An IPNS record is a data structure composed by:

- 1. **Value** (bytes)
- 2. **Validity** (bytes)
- 3. **Validity Type** (uint64)
  - Allows us to define if the record is valid under certain conditions
  - Only supporting expiration date with `validityType = 0` for now
- 4. **Signature** (bytes)
  - Concatenate value, validity field and validity type

>   - Sign the concatenation result with the provided private key

Author: Yes, totally agree, we need to change the signature logic! 

We should rethink this once we make changes to the record definition.

  
These records are stored locally, as well as spread accross the network, in order to be accessible to everyone. For storing this structured data, we use [Protocol Buffers](https://github.com/google/protobuf), which is a language-neutral, platform neutral extensible mechanism for serializing structured data.

```
message IpnsEntry {
	enum ValidityType {
		// setting an EOL says "this record is valid until..."
		EOL = 0;
	}
	required bytes value = 1;
	required bytes signature = 2;

	optional ValidityType validityType = 3;
	optional bytes validity = 4;

	optional uint64 sequence = 5;

	optional uint64 ttl = 6;

	optional bytes pubKey = 7;
}
```

The record is stored locally, in order to allow us to understand which is the most recent record in the network, since it will be stored locally by the peer using the keys used to generate it.

**Note**: validity must be with nanoseconds precision.

#### Local record

This record is stored in the peer's repo datastore and contains the **latest** version of the IPNS record published by the provided key. This record is useful for republishing, as well as tracking the sequence number.

**Key format:** `/ipns/base32(<HASH>)`

> Note: Base32 according to the [RFC4648](https://tools.ietf.org/html/rfc4648).

Author: I will move this for now to the bottom of the document. We will see afterward if it makes sense to remove it!
